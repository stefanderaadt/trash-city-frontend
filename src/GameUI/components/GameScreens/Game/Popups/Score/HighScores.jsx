import React from 'react';

import Button from '../../../../Default/Button';
import TextInput from '../../../../Default/TextInput';
import displayTime from '../../../../../../utils/displayTime';

const HighScores = ({
  setScoreName,
  saveHighScore,
  score: { scoreName, highScores, fetchingHighScores, fetchedHighScores, savingScore, savedScore },
  time
}) => {
  if (fetchingHighScores) {
    return <i className="icon-busy icon-hour-glass" />;
  }

  if (!fetchedHighScores) {
    return null;
  }

  return (
    <div className="score-popup-scores">
      <div className="score-popup-highscores">
        <h2>Beste tijden</h2>
        {highScores.map((highScore, key) => (
          <p key={highScore.id}>
            {`${key + 1}: ${highScore.name} - ${displayTime(highScore.score)}`}
          </p>
        ))}
      </div>
      {(highScores.length < 5 || time < highScores[highScores.length - 1].score) && (
        <div className="score-popup-save">
          <p>Je hebt een top 5 tijd!</p>
          <TextInput
            className="popup-score-name-input"
            type="text"
            placeholder="Naam"
            value={scoreName}
            onChange={e => setScoreName(e.target.value)}
          />
          <Button
            className="popup-button-save-highscore"
            onClick={saveHighScore}
            busy={savingScore}
            disabled={savedScore}
          >
            Score opslaan
          </Button>
        </div>
      )}
    </div>
  );
};

export default HighScores;
