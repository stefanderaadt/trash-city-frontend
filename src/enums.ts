enum Screens {
  Loading,
  MainMenu,
  LevelMenu,
  SettingsMenu,
  ControlsMenu,
  Game
}

enum Popups {
  None,
  Pause,
  Score,
  Info
}

enum GameEventIds {
  SidewalkArea = 1,
  CivilianSpawnPoint = 2,
  WaterTile = 3,
  FilterSystem = 4,
  CleanerSpawnPoint = 5
}

enum TileLayers {
  Background,
  TopAssets1,
  TopAssets2
}

enum ObjectDirection {
  Up,
  Down,
  Left,
  Right
}

enum Keys {
  ArrowUp = 38,
  ArrowDown = 40,
  ArrowLeft = 37,
  ArrowRight = 39,
  SpaceBar = 32
}

enum Flags {
  FlippedHorizontally = 0x80000000,
  FlippedVertically = 0x40000000,
  FlippedDiagonally = 0x20000000
}

export { Screens, Popups, GameEventIds, ObjectDirection, Keys, TileLayers, Flags };
