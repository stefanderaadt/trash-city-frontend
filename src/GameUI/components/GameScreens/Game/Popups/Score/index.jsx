import React from 'react';

import GamePopup from '../GamePopup';
import Button from '../../../../Default/Button';
import { Screens } from '../../../../../../enums.ts';
import displayTime from '../../../../../../utils/displayTime';
import HighScores from './HighScores';

const Score = ({ game, game: { time }, setScreen, setScoreName, saveHighScore, score }) => (
  <GamePopup className="score-popup">
    <div className="score-popup-content">
      <h1>Gewonnen!</h1>
      <hr />
      <div className="score-popup-info">
        <p>{`Tijd: ${displayTime(game.time)}`}</p>
      </div>
      <hr />
      <HighScores
        score={score}
        time={time}
        saveHighScore={saveHighScore}
        setScoreName={setScoreName}
      />
      <hr />
      <div className="popup-button-container">
        <Button className="popup-button-mainmenu" onClick={() => setScreen(Screens.MainMenu)}>
          Hoofdmenu
        </Button>
      </div>
    </div>
  </GamePopup>
);

export default Score;
