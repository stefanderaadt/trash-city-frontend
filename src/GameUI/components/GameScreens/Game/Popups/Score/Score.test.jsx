/* eslint-disable no-console */
import React from 'react';
import { mount, render } from 'enzyme';

import { Screens } from '../../../../../../enums.ts';
import Score from '.';

jest.mock('../../../../../../Game/Game.ts', () => {
  return {};
});

describe('Score popup tests', () => {
  test('Score mainmenu button working as expected', () => {
    let screen;
    const wrapper = mount(
      <Score
        game={{
          score: 0,
          time: 300
        }}
        setScreen={s => {
          screen = s;
        }}
        score={{
          scoreName: 'test',
          highScores: [
            { id: '1245215215g', name: 'Stefan', score: 182 },
            { id: '12455215g', name: 'Naam', score: 12 }
          ],
          fetchingHighScores: false,
          fetchedHighScores: true,
          savingScore: false,
          savedScore: false
        }}
      />
    );
    const button = wrapper.find('.popup-button-mainmenu').first();
    button.simulate('click');
    expect(screen).toEqual(Screens.MainMenu);
  });

  test('Score mainmenu button working as expected', () => {
    let name;
    const wrapper = mount(
      <Score
        game={{
          score: 0,
          time: 300
        }}
        score={{
          scoreName: 'not the right name',
          highScores: [
            { id: '1245215215g', name: 'Stefan', score: 182 },
            { id: '12455215g', name: 'Naam', score: 12 }
          ],
          fetchingHighScores: false,
          fetchedHighScores: true,
          savingScore: false,
          savedScore: false
        }}
        setScoreName={n => {
          name = n;
        }}
      />
    );
    const input = wrapper.find('.popup-score-name-input').first();
    input.simulate('change', { target: { value: 'testname' } });
    expect(name).toEqual('testname');
  });

  test('Score rendering correctly with fetchingHighScores = true', () => {
    const wrapper = render(
      <Score
        game={{
          score: 0,
          time: 300
        }}
        score={{
          scoreName: 'nameeee',
          highScores: [
            { id: '1245215215g', name: 'Stefan', score: 182 },
            { id: '12455215g', name: 'Naam', score: 12 }
          ],
          fetchingHighScores: true,
          fetchedHighScores: false,
          savingScore: false,
          savedScore: false
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Score rendering correctly with fetchedHighScores = false', () => {
    const wrapper = render(
      <Score
        game={{
          score: 0,
          time: 300
        }}
        score={{
          scoreName: 'nameeee',
          highScores: [
            { id: '1245215215g', name: 'Stefan', score: 182 },
            { id: '12455215g', name: 'Naam', score: 12 }
          ],
          fetchingHighScores: false,
          fetchedHighScores: false,
          savingScore: false,
          savedScore: false
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
