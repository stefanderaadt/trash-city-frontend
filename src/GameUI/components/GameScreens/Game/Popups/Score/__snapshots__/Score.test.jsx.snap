// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score popup tests Score rendering correctly with fetchedHighScores = false 1`] = `
<div
  class="game-popup score-popup"
>
  <div
    class="game-popup-container"
  >
    <div
      class="score-popup-content"
    >
      <h1>
        Gewonnen!
      </h1>
      <hr />
      <div
        class="score-popup-info"
      >
        <p>
          Tijd: 05:00
        </p>
      </div>
      <hr />
      <hr />
      <div
        class="popup-button-container"
      >
        <div
          class="button popup-button-mainmenu"
          role="button"
          tabindex="0"
        >
          Hoofdmenu
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Score popup tests Score rendering correctly with fetchingHighScores = true 1`] = `
<div
  class="game-popup score-popup"
>
  <div
    class="game-popup-container"
  >
    <div
      class="score-popup-content"
    >
      <h1>
        Gewonnen!
      </h1>
      <hr />
      <div
        class="score-popup-info"
      >
        <p>
          Tijd: 05:00
        </p>
      </div>
      <hr />
      <i
        class="icon-busy icon-hour-glass"
      />
      <hr />
      <div
        class="popup-button-container"
      >
        <div
          class="button popup-button-mainmenu"
          role="button"
          tabindex="0"
        >
          Hoofdmenu
        </div>
      </div>
    </div>
  </div>
</div>
`;
